### PR TITLE
feat: [PL-41387]: OIDC Token Library implementation for ID Token generation

### PR DESCRIPTION
**Pull request by Raghav-Murali**
## Describe your changes

This PR implements the ID Token generation part for OIDC Token library.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/54211)

### CODE REVIEW

**Comment by Raghav-Murali** [comment]
trigger uts

**Comment by Raghav-Murali** [comment]
trigger uts

**Comment by Raghav-Murali** [comment]
trigger uts

**Comment by Raghav-Murali** [comment]
trigger ut6

File: 930-oidc-core/src/main/java/io/harness/oidc/idtoken/OidcIdTokenConstants.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  public static final String TYPE = "typ";
+  public static final String ALGORITHM = "alg";
+  public static final String KEY_IDENTIFIER = "kid";
+
+  public static final String SUBJECT = "sub";
+  public static final String AUDIENCE = "aud";
+  public static final String ISSUER = "iss";
+  public static final String EXPIRY = "exp";
+  public static final String ISSUED_AT = "iat";
+  public static final String ACCOUNT_ID = "account_id";
```
Comment (gn-harness) [inline 930-oidc-core/src/main/java/io/harness/oidc/idtoken/OidcIdTokenConstants.java:24]: I would have added organization and pipeline information from day1. We already know customer needs it and its not much additional work too. CI team would have to later again create new PR for passing pipeline information too.

not a blocker comment. Lets do it as a quick followup

**Review by gn-harness — APPROVED** [review]

File: 930-oidc-core/src/main/java/io/harness/oidc/idtoken/OidcIdTokenUtility.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+import java.util.Base64;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import lombok.extern.slf4j.Slf4j;
+
+@OwnedBy(HarnessTeam.PL)
+@Slf4j
+public class OidcIdTokenUtility {
```
[the remark this answers is not in the document]
Comment (Raghav-Murali) [reply 930-oidc-core/src/main/java/io/harness/oidc/idtoken/OidcIdTokenUtility.java:32]: Done

File: 930-oidc-core/src/main/java/io/harness/oidc/idtoken/OidcIdTokenUtility.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+      // Get the Payload object into a Map
+      Map<String, Object> payloadMap = objectMapper.convertValue(oidcIdTokenPayloadStructure, Map.class);
+
+      // Load the RSA private key from the String
+      PrivateKey privateKey = loadPrivateKeyFromString(rsaPrivateKeyPEM);
+
+      // Build the JWT using the private key
+      String jwt = Jwts.builder()
+                       .setHeader(headerMap)
```
[the remark this answers is not in the document]
Comment (Raghav-Murali) [reply 930-oidc-core/src/main/java/io/harness/oidc/idtoken/OidcIdTokenUtility.java:62]: Will not take it as an arg.

**Pull request merged**